src/file.cpp

```diff
@@ -90,16 +92,73 @@ file file::copy(const fs::path& path, std::string_view label,
 }
 
 std::string file::read() const {
-  std::ifstream stream = input_stream();
-  return std::string(std::istreambuf_iterator<char>(stream), {});
+  native_handle_type handle = native_handle();
+
+#ifdef _WIN32
+  SetFilePointer(handle, 0, nullptr, FILE_BEGIN);
+#else
+  lseek(handle, 0, SEEK_SET);
+#endif
+
+  std::string content;
+  content.resize(file_size(path()));
+
+  std::size_t offset = 0;
+  while (offset < content.size()) {
+#ifdef _WIN32
+    DWORD bytes_read;
+    BOOL ret = ReadFile(handle, &content[offset], content.size() - offset, &bytes_read, nullptr);
+    if (!ret) {
+      std::error_code ec = std::error_code(GetLastError(), std::system_category());
+      throw fs::filesystem_error("Cannot read a temporary file", path(), ec);
+    }
+#else
+    ssize_t bytes_read = ::read(handle, &content[offset], content.size() - offset);
+    if (bytes_read < 0) {
+      std::error_code ec = std::error_code(errno, std::system_category());
+      throw fs::filesystem_error("Cannot read a temporary file", path(), ec);
+    }
+#endif
+
+    offset += bytes_read;
+  }
+
+  return content;
 }
 
 void file::write(std::string_view content) const {
-  output_stream(std::ios::trunc) << content;
+  fs::resize_file(path(), 0);
+  append(content);
 }
 
 void file::append(std::string_view content) const {
-  output_stream(std::ios::app) << content;
+  native_handle_type handle = native_handle();
+
+#ifdef _WIN32
+  SetFilePointer(handle, 0, nullptr, FILE_END);
+#else
+  lseek(handle, 0, SEEK_END);
+#endif
+
+  std::size_t offset = 0;
```

Comment:
You can modify passed `content` string (or make a modifiable copy of it, if that would feel clearer) instead of manually tracking a separate offset variable.

Always using a string view without pointer arithmetic would be less bugprone, I believe